### PR TITLE
Adds Internet admin midi

### PR DESCRIPTION
**Pull request by JJRcop**
:cl: JJRcop
add: Admins can now play media content from the web to players.
/:cl:

This exposes https://github.com/rg3/youtube-dl to admins for playing music to players. Admins may provide a content URL, which gets passed to youtube-dl to retrieve a raw audio URL which is then sent to clients and played within the goonchat control. This respects the admin midi pref.

Also added a volume slider to the goonchat options that allows players to change the volume of playing web content, this also affects Play Global Sound.

_Youtube-dl supports a large amount of websites, you can play content from: YouTube, Soundcloud, Bandcamp, and more. This does depend on if the website provides audio content compatible with IE11, which, for example, twitch.tv does not provide, even though twitch.tv is supported by youtube-dl._

_For hosts: This does not install youtube-dl for you, you must install it yourself and then edit the INVOKE_YOUTUBEDL option in config.txt appropriately, or just uncomment it if youtube-dl is in your PATH_

### Demonstration video: https://www.youtube.com/watch?v=_IyhmEfpAVQ

### CODE REVIEW

File: code/__HELPERS/shell.dm [hunk 1]
```diff
@@ -0,0 +1,56 @@
+//Runs the command in the system's shell, returns a list of (error code, stdout, stderr)
```
Comment (JJRcop) [inline code/__HELPERS/shell.dm]: This file deals with the ```shell()``` proc, thus it has the potential to carry a shell injection exploit. I've done my best, but please be extra thorough when reviewing the code here.

**Comment by pubby** [comment]
youtube-dl gets updated a ton because the websites it scrapes are constantly changing

It might be better to make it a git submodule or figure out how to make it auto-update so that it doesn't break every X months

**Comment by JJRcop** [comment]
@pubby #28871

File: code/__HELPERS/shell.dm [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+#define SHELLEO_NAME "data/shelleo."
+#define SHELLEO_ERR ".err"
+#define SHELLEO_OUT ".out"
+/proc/shelleo(command)
```
Comment (Cyberboss) [inline code/__HELPERS/shell.dm]: /world/proc

/proc can be admin called and abused

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/__HELPERS/shell.dm (hunk 2 above)
Comment (JJRcop) [reply code/__HELPERS/shell.dm]: yikes!

**Comment by Cyberboss** [comment]
@tgstation/commit-access Submodule? Or are our contributors too _special_ to handle it?

**Comment by JJRcop** [comment]
I would love to make it a submodule instead if possible

**Comment by Repukan** [comment]
Can you strip kevinz000 of sound+ first.

**Comment by GunHog** [comment]
I was going to make a joke like that, Repukan! Poor @kevinz000

**Comment by PKPenguin321** [comment]
This is one of those features that I've dreamed of but figured it would be too impossible for anyone to code
Bravo

**Comment by ike709** [comment]
871 files changed and 137,319 lines added just so admins can play shitty meme music.

SS13 in a nutshell.

**Comment by RemieRichards** [comment]
This absolutely needs to be outside of our repo (read: submodule)

**Comment by AnturK** [comment]
Yep. External if at all.

**Comment by MrStonedOne** [comment]
Wow, ya, thats alot of data.

I guess submodule it is. @Cyberboss could you make tgstation-server fetch and update submodules?

In this case, coders and dev boxes don't really need this, so not fetching the submodule by default wouldn't hurt anything in their case.

@JJRcop you might also have to roll back the commits so those files don't eat up the repo, but a squish and merge would also do the trick.

**Comment by MrStonedOne** [comment]
@JJRcop you could also config gate this, and then assume youtube-dl will be in path or some known folder and require the user handle putting it there/updating it, if submodule doesn't end up working out.

**Comment by Dorsisdwarf** [comment]
I like the implentation, as well as the fact you're finally giving players a way to tune their enjoyment (or lack thereof) of admin sounds other than "mute or MAX VOLUME"

**Comment by JJRcop** [comment]
Only two out of nine maintainers outright gave an answer, even though the host also supports submodules.
@Jordie0608 @lzimann @KorPhaeron @ChangelingRain @Cheridan @optimumtact @Cyberboss (cyberboss called for opinions but didn't give his own yet)

**Comment by Cyberboss** [comment]
I'm against purely because people kept deleting it last time and then
asking how to fix it in their PRs. But, this really is a textbook use case
so I'm not going to argue. May reopen my fixer script PR though.

**Comment by JJRcop** [comment]
LeoZ (to avoid pinging him again) said in IRC that he doesn't care.

**Comment by AnturK** [comment]
We could just move it to additional dependencies install in server tools and just check if it exists. It's not really crucial feature after all and can be left for host to configure.

**Comment by Tacolizard** [comment]
:thinking:

we are messing with power we barely understand. Caution is advised

**Comment by JJRcop** [comment]
For the moment I will rebase to remove the youtube-dl source code and just assume it's in your path, pending what we do next.

**Comment by JJRcop** [comment]
here we go again

File: code/__HELPERS/shell.dm [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		shelleo_ids[shelleo_id] = FALSE
+	else
+		CRASH("Operating System: [world.system_type] not supported") // If you encounter this error, you are encouraged to update this proc with support for the new operating system
+	. = list(errorcode, stdout, stderr)
+#undef SHELLEO_NAME
+#undef SHELLEO_ERR
+#undef SHELLEO_OUT
+
+/proc/shell_url_scrub(url)
+	var/static/regex/bad_chars_regex = regex("\[^#%&./:=?\\w]*", "g")
```
Comment (JJRcop) [inline code/__HELPERS/shell.dm]: This regex means "not any of these: #%&./:=?\w" (\w is [a-z][A-Z][0-9]_)

**Comment by Cyberboss** [comment]
Need to do tweaking in the server tools before this is merged

**Comment by JJRcop** [comment]
@Cyberboss How is the code here besides that?

File: code/modules/admin/verbs/playsound.dm [hunk 4]
```diff
@@ -52,6 +56,51 @@
...
+					web_sound_url = html_url_regex.match
+
+					if(SSevents.holidays && SSevents.holidays[APRIL_FOOLS])
+						pitch = pick(0.5, 0.7, 0.8, 0.85, 0.9, 0.95, 1.1, 1.2, 1.4, 1.6, 2.0, 2.5)
+						to_chat(src, "You feel the Honkmother messing with your song...")
+
+					log_admin("[key_name(src)] played web sound: [web_sound_input]")
+					message_admins("[key_name(src)] played web sound: [web_sound_input]")
+			else
+				to_chat(src, "<span class='boldwarning'>Youtube-dl URL retrieval FAILED:\n<span class='warning'>[stderr]</span></span>")
```
Comment (Cyberboss) [inline code/modules/admin/verbs/playsound.dm]: use multiple to_chats instead of \n

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by JJRcop** [comment]
Can someone remove the do not merge tag?

**Comment by Cyberboss** [comment]
It still needs the other two server tools prs to be merged before I can
make the new release

**Comment by Iamgoofball** [comment]
:-1:

``<iframe width="0" height="0" src="https://www.youtube.com/embed/youtube_id_here?autoplay=1" frameborder="0"></iframe>``

boom

no youtube downloader garbage

just update byond and update IE and it works

**Comment by Cyberboss** [comment]
We lose a lot of finesse that we have with this if we do that

**Comment by MrStonedOne** [comment]
iamgoofball. this systems generates a audio only stream, no need to make clients download videos.

**Comment by optimumtact** [comment]
Do not use submodule. Expose config to set path for youtube dl and leave serveradmin to determine its install ie via package manager or git or etc

**Comment by optimumtact** [comment]
Submoduoe too annoying plus i dont want two installs of youtube dl on a system just because i run byond as well

**Comment by optimumtact** [comment]
I see anturk plus mso beat me too it

**Comment by optimumtact** [comment]
Plus we avoid potential licensing issues

**Comment by JJRcop** [comment]
For the licensing, youtube-dl has a [LICENSE](https://github.com/rg3/youtube-dl/blob/5113b6912467619bd463c5ebefe759d07078bea1/LICENSE) which declares it under public domain. Your other points are still valid, though.

**Comment by optimumtact** [comment]
The last submodule we had got deleted at least twice

**Comment by JJRcop** [comment]
With that in mind, if we have problems again, I am switching to the config idea very rapidly.

**Comment by JJRcop** [comment]
I counted the votes wrong, 1 for, 3 against, 2 abstain. Switching to config option

File: config/config.txt [hunk 5]
```diff
@@ -195,6 +195,14 @@ CHECK_RANDOMIZER
 ## Ban appeals URL - usually for a forum or wherever people should go to contact your admins.
 # BANAPPEALS http://justanotherday.example.com
 
+## System command that invokes youtube-dl, used by Play Internet Sound.
+## You can install youtube-dl with
+## "pip install youtube-dl" if you have pip installed
+## from https://github.com/rg3/youtube-dl/releases
+## or your package manager
+## The default value assumes youtube-dl is in your system PATH
```
Comment (JJRcop) [inline config/config.txt]: For a moment I thought I won the "longest config option description" award, but USE_AGE_RESTRICTION_FOR_JOBS still claims that prize.

File: code/controllers/configuration.dm [hunk 6]
```diff
@@ -12,7 +12,7 @@
 	return ..()
 
 /datum/configuration/vv_edit_var(var_name, var_value)
-	var/static/list/banned_edits = list("cross_address", "cross_allowed", "autoadmin", "autoadmin_rank")
+	var/static/list/banned_edits = list("cross_address", "cross_allowed", "autoadmin", "autoadmin_rank", "invoke_youtubedl")
```
Comment (JJRcop) [inline code/controllers/configuration.dm]: Is there anywhere else I need to place this to prevent tampering?

Comment (Cyberboss) [reply code/controllers/configuration.dm]: no

**Comment by JJRcop** [comment]
All ready to go, please review this again since I rebased it.

**Comment by Iamgoofball** [comment]
s t o p

use iframes

christ

**Comment by JJRcop** [comment]
I initially wanted to use youtube and soundcloud's embed, but after a [discussion](https://file.house/JunY.txt) I decided to go with youtube-dl, which supports both youtube, soundcloud, and many more sites than I could think of.

**Comment by optimumtact** [comment]
@Iamgoofball stop suggesting options that are worse you muppet

File: code/modules/admin/admin_verbs.dm [hunk 7]
```diff
@@ -74,7 +74,7 @@ GLOBAL_LIST_INIT(admin_verbs_admin, world.AVerbsAdmin())
 GLOBAL_PROTECT(admin_verbs_ban)
 GLOBAL_LIST_INIT(admin_verbs_ban, list(/client/proc/unban_panel, /client/proc/DB_ban_panel, /client/proc/stickybanpanel))
 GLOBAL_PROTECT(admin_verbs_sounds)
-GLOBAL_LIST_INIT(admin_verbs_sounds, list(/client/proc/play_local_sound, /client/proc/play_sound, /client/proc/set_round_end_sound))
+GLOBAL_LIST_INIT(admin_verbs_sounds, list(/client/proc/play_web_sound, /client/proc/play_local_sound, /client/proc/play_sound, /client/proc/set_round_end_sound))
```
Comment (Cyberboss) [inline code/modules/admin/admin_verbs.dm]: They shouldn't get the verb if youtube-dl isn't installed

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/admin/admin_verbs.dm (hunk 7 above)
Comment (JJRcop) [reply code/modules/admin/admin_verbs.dm]: 👌

File: code/modules/admin/verbs/playsound.dm [hunk 8]
```diff
@@ -52,6 +56,56 @@
...
+		to_chat(src, "<span class='boldwarning'>Youtube-dl was not configured, action unavailable</span>") //Check config.txt for the INVOKE_YOUTUBEDL value
+		return
+
+	var/web_sound_input = input("Enter content URL (supported sites only, leave blank to stop playing)", "Play Internet Sound via youtube-dl") as text|null
+	if(istext(web_sound_input))
+		var/web_sound_url = ""
+		var/pitch
+		if(length(web_sound_input))
+
+			web_sound_input = trim(web_sound_input)
```
Comment (Cyberboss) [inline code/modules/admin/verbs/playsound.dm]: trimming should be done before checking length

File: code/modules/admin/verbs/playsound.dm [hunk 9]
```diff
@@ -52,6 +56,56 @@
...
+			var/shell_scrubbed_input = shell_url_scrub(web_sound_input)
+			var/list/output = world.shelleo("[config.invoke_youtubedl] --format \"bestaudio\[ext=aac]/bestaudio\[ext=mp3]/bestaudio\[ext=m4a]\" --get-url \"[shell_scrubbed_input]\"")
+			var/errorlevel = output[1]
+			var/stdout = output[2]
+			var/stderr = output[3]
```
Comment (Cyberboss) [inline code/modules/admin/verbs/playsound.dm]: These numbers should be defines

File: code/modules/admin/verbs/playsound.dm [hunk 10]
```diff
@@ -52,6 +56,56 @@
...
+				to_chat(src, "<span class='boldwarning'>Youtube-dl URL retrieval FAILED:</span>")
+				to_chat(src, "<span class='warning'>[stderr]</span>")
+
+		else //pressed ok with blank
+			log_admin("[key_name(src)] stopped web sound")
+			message_admins("[key_name(src)] stopped web sound")
+			web_sound_url = " "
+
+		if(web_sound_url)
+			for(var/mob/M in GLOB.player_list)
```
Comment (Cyberboss) [inline code/modules/admin/verbs/playsound.dm]: typeless loop

File: code/modules/admin/verbs/playsound.dm [hunk 11]
```diff
@@ -52,6 +56,56 @@
...
+				var/client/C = M.client
+				if(C.prefs.toggles & SOUND_MIDI && C.chatOutput && !C.chatOutput.broken && C.chatOutput.loaded)
```
Comment (Cyberboss) [inline code/modules/admin/verbs/playsound.dm]: brackets around the & pleae

**Review by Cyberboss — DISMISSED** [review]

File: code/modules/admin/verbs/playsound.dm (hunk 8 above)
Comment (JJRcop) [reply code/modules/admin/verbs/playsound.dm]: I did this on purpose so that ONLY if the input is completely blank will it stop audio

**Review by optimumtact — APPROVED** [review]

File: code/modules/admin/verbs/playsound.dm [hunk 12]
```diff
@@ -52,6 +56,57 @@
...
+		return
+
+	var/web_sound_input = input("Enter content URL (supported sites only, leave blank to stop playing)", "Play Internet Sound via youtube-dl") as text|null
+	if(istext(web_sound_input))
+		var/web_sound_url = ""
+		var/pitch
+		if(length(web_sound_input))
+
+			web_sound_input = trim(web_sound_input)
+			var/shell_scrubbed_input = shell_url_scrub(web_sound_input)
```
Comment (MrStonedOne) [inline code/modules/admin/verbs/playsound.dm]: for just an extra and likely unneeded layer of security (yay) add a `url_encode()` here

`var/shell_scrubbed_input = shell_url_scrub(url_encode(web_sound_input, TRUE))`

File: code/__HELPERS/shell.dm [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	var/scrubbed_url = ""
+	var/bad_match = ""
+	var/last_good = 1
+	var/bad_chars = 1
+	while(bad_chars)
```
Comment (MrStonedOne) [inline code/__HELPERS/shell.dm]: You could convert this to a do while loop since thats what it is.

File: code/modules/admin/verbs/playsound.dm [hunk 14]
```diff
@@ -73,15 +73,20 @@
 		if(length(web_sound_input))
 
 			web_sound_input = trim(web_sound_input)
+			var/static/regex/html_protocol_regex = regex("https?://")
+			if(findtext(web_sound_input, ":") && !findtext(web_sound_input, html_protocol_regex))
+				to_chat(src, "<span class='boldwarning'>Non-http(s) URIs are not allowed.</span>")
+				to_chat(src, "<span class='warning'>For youtube-dl shortcuts like ytsearch: please use the appropriate full url from the website.</span>")
```
Comment (MrStonedOne) [inline code/modules/admin/verbs/playsound.dm]: It occurs to me you could manually allow these, since youtube would obviously parse them

Comment (JJRcop) [reply code/modules/admin/verbs/playsound.dm]: There are a lot more websites than youtube supported by youtube-dl, I don't want to whitelist every one.

**Review by Cyberboss — CHANGES_REQUESTED** [review]
do while

**Review by Cyberboss — APPROVED** [review]

**Comment by Iamgoofball** [comment]
not even 5 hours and we can't autorevert it

**Comment by JJRcop** [comment]
@Iamgoofball I see a revert button next to cyberboss' merge commit listing

**Pull request merged**